man/man8/vlogconf.8
.TH vlogconf 8 "March 2008" "OpenFlow" "OpenFlow Manual"

.SH NAME
vlogconf \- configuration utility for OpenFlow logging in userspace

.SH SYNOPSIS
\fBvlogconf\fR [\fB-h\fR | \fB--help\fR] [TARGET...] [ACTION...]
.sp 1
The available TARGET options are:
.br
[\fB-a\fR | \fB--all\fR] [\fB-t\fR PID | \fB--target=\fIPID\fR]
.sp 1
The available ACTION options are:
.br
[\fB-l\fR | \fB--list\fR] [\fB-s\fR
\fIMODULE\fB:\fIFACILITY\fB:\fILEVEL\fR |
\fB--set=\fIMODULE\fB:\fIFACILITY\fB:\fILEVEL\fR]

.SH DESCRIPTION
The \fBvlogconf\fR program configures the logging system used by the 
OpenFlow userspace programs.  The logging configuration may be modified 
while OpenFlow programs are running.

\fBvlogconf\fR applies one or more actions to each of one or more
target processes.  Targets may be specified as:

.TP
\fB-a\fR, \fB--all\fR
All running processes that \fBvlogconf\fR can control.

.TP
\fB-t \fIPID\fR, \fB--target=\fIPID\fR
The process with the specified \fIPID\fR.  \fIPID\fR may also specify
an absolute path (beginning with `/') to the Unix domain socket for a
\fBvlogconf\fR-controllable process.

.PP
The available actions are:

.TP
\fB-l\fR, \fB--list\fR
Print the list of known modules and their current logging levels to
stdout.

.TP
\fB-s\fR \fIMODULE\fB:\fIFACILITY\fB:\fILEVEL\fR, \fB--set=\fIMODULE\fB:\fIFACILITY\fB:\fILEVEL\fR

Sets the logging level for \fIMODULE\fR in \fIFACILITY\fR to
\fILEVEL\fR.  \fIMODULE\fR may be any valid module name (as displayed
by the \fB--list\fR option) or the special name \fBANY\fR to set all
modules logging level.  \fIFACILITY\fR may be \fBsyslog\fR or
\fBconsole\fR to set the levels for logging to the system log or to
the console, respectively, or \fBANY\fR to set the logging levels for
both modules.  \fILEVEL\fR must be one of \fBemer\fR, \fBerr\fR,
\fBwarn\fR, or \fBdbg\fR, designating the minimum severity of a
message for it to be logged.

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\^\-help\fR
Prints a brief help message to the console and exits.


.SH "SEE ALSO"

.BR dpctl (8),
.BR secchan (8),
.BR controller (8)


utilities/vlogconf.c
#include "vlog.h"

#include <dirent.h>
#include <errno.h>
#include <getopt.h>
#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "command-line.h"
#include "compiler.h"
#include "util.h"
#include "vlog-socket.h"

void
usage(char *prog_name, int exit_code)
{
    printf("Usage: %s [TARGET] [ACTION...]\n"
           "Targets:\n"
           "  -a, --all            Apply to all targets (default)\n"
           "  -t, --target=TARGET  Specify target program, as a pid or an\n"
           "                       absolute path to a Unix domain socket\n"
           "Actions:\n"
           "  -l, --list         List current settings\n"
           "  -s, --set=MODULE:FACILITY:LEVEL\n"
           "        Set MODULE and FACILITY log level to LEVEL\n"
           "        MODULE may be any valid module name or 'ANY'\n"
           "        FACILITY may be 'syslog' or 'console' or 'ANY'\n"
           "        LEVEL may be 'emer', 'err', 'warn', or 'dbg'\n"
           "  -h, --help         Print this helpful information\n",
           prog_name);
    exit(exit_code);
}

static char *
transact(struct vlog_client *client, const char *request, bool *ok)
{
    char *reply;
    int error = vlog_client_transact(client, request, &reply);
    if (error) {
        fprintf(stderr, "%s: transaction error: %s\n",
                vlog_client_target(client), strerror(error));
        *ok = false;
    }
    return reply ? reply : xstrdup("");
}

static void
transact_ack(struct vlog_client *client, const char* request, bool *ok)
{
    char *reply;
    int error = vlog_client_transact(client, request, &reply);
    if (error) {
        fprintf(stderr, "%s: transaction error: %s\n",
                vlog_client_target(client), strerror(error));
        *ok = false;
    } else if (strcmp(reply, "ack")) {
        fprintf(stderr, "Received unexpected reply from %s: %s\n",
                vlog_client_target(client), reply);
        *ok = false;
    }
    free(reply);
}

static void
add_target(struct vlog_client ***clients, size_t *n_clients,
           const char *path, bool *ok)
{
    struct vlog_client *client;
    int error = vlog_client_connect(path, &client);
    if (error) {
        fprintf(stderr, "Error connecting to \"%s\": %s\n",
                path, strerror(error));
        *ok = false;
    } else {
        *clients = xrealloc(*clients, sizeof *clients * (*n_clients + 1));
        (*clients)[*n_clients] = client;
        ++*n_clients;
    }
}

static void
add_all_targets(struct vlog_client ***clients, size_t *n_clients, bool *ok)
{
    DIR *directory;
    struct dirent* de;

    directory = opendir("/tmp");
    if (!directory) {
        fprintf(stderr, "/tmp: opendir: %s\n", strerror(errno));
    }

    while ((de = readdir(directory)) != NULL) {
        if (!strncmp(de->d_name, "vlogs.", 5)) {
            char *path = xasprintf("/tmp/%s", de->d_name);
            add_target(clients, n_clients, path, ok);
            free(path);
        }
    }

    closedir(directory);
}

int main(int argc, char *argv[])
{
    static const struct option long_options[] = {
        /* Target options must come first. */
        {"all", no_argument, NULL, 'a'},
        {"target", required_argument, NULL, 't'},
        {"help", no_argument, NULL, 'h'},

        /* Action options come afterward. */
        {"list", no_argument, NULL, 'l'},
        {"set", required_argument, NULL, 's'},
        {0, 0, 0, 0},
    };
    char *short_options;

    /* Determine targets. */
    bool ok = true;
    int n_actions = 0;
    struct vlog_client **clients = NULL;
    size_t n_clients = 0;

    set_program_name(argv[0]);

    short_options = long_options_to_short_options(long_options);
    for (;;) {
        int option;
        size_t i;

        option = getopt_long(argc, argv, short_options, long_options, NULL);
        if (option == -1) {
            break;
        }
        if (!strchr("ath", option) && n_clients == 0) {
            fatal(0, "no targets specified (use --help for help)");
        } else {
            ++n_actions;
        }
        switch (option) {
        case 'a':
            add_all_targets(&clients, &n_clients, &ok);
            break;

        case 't':
            add_target(&clients, &n_clients, optarg, &ok);
            break;

        case 'l':
            for (i = 0; i < n_clients; i++) {
                struct vlog_client *client = clients[i];
                char *reply;

                printf("%s:\n", vlog_client_target(client));
                reply = transact(client, "list", &ok);
                fputs(reply, stdout);
                free(reply);
            }
            break;

        case 's':
            for (i = 0; i < n_clients; i++) {
                struct vlog_client *client = clients[i];
                char *request = xasprintf("set %s", optarg);
                transact_ack(client, request, &ok);
                free(request);
            }
            break;

        case 'h':
            usage(argv[0], EXIT_SUCCESS);
            break;

        default:
            NOT_REACHED();
        }
    }
    if (!n_actions) {
        fprintf(stderr,
                "warning: no actions specified (use --help for help)\n");
    }
    exit(ok ? 0 : 1);
}
